libraries/lib-cloud-audiocom/UploadService.cpp

```diff
@@ -0,0 +1,475 @@
+              mServiceConfig.GetFinishUploadPage(mAudioID, mUploadToken),
+              mAudioSlug });
+      }
+      
+      mProgressCallback = {};
+      mCompletedCallback = {};
+   }
+
+   void InitiateUpload(std::string_view authToken)
+   {      
+      using namespace audacity::network_manager;
+
+      Request request(mServiceConfig.GetAPIUrl("/audio"));
+
+      request.setHeader("Content-Type", "application/json");
+
+      mAuthToken = std::string(authToken);
+      SetAuthHeader(request);
+
+      const auto payload = GetUploadRequestPayload(mFileName, mProjectName);
+
+      std::lock_guard<std::mutex> lock(mStatusMutex);
+
+      // User has already aborted? Do not send the request.
+      if (mAborted)
+         return;
+
+      auto response = NetworkManager::GetInstance().doPost(
+         request, payload.data(), payload.size());
+
+      mActiveResponse = response;
+
+      response->setRequestFinishedCallback(
+         [response, sharedThis = shared_from_this(), this](auto) {
+            auto responseCode = response->getHTTPCode();
+            
+            if (responseCode == 201)
+            {
+               HandleUploadPolicy(response->readAll<std::string>());
+            }
+            else if (responseCode == 401)
+            {
+               FailPromise(
+                  UploadOperationCompleted::Result::Unauthorized,
+                  response->readAll<std::string>());
+            }
+            else if (responseCode == 422)
+            {
+               FailPromise(
+                  UploadOperationCompleted::Result::InvalidData,
+                  response->readAll<std::string>());
+            }
+            else
+            {
+               FailPromise(
+                  UploadOperationCompleted::Result::UnexpetedResponse,
+                  response->readAll<std::string>());
+            }
+         });
+   }
+
+   void HandleUploadPolicy(std::string uploadPolicyJSON)
+   {
+      using namespace audacity::network_manager;
+      
+      rapidjson::Document document;
+      document.Parse(uploadPolicyJSON.data(), uploadPolicyJSON.length());
+
+      if (
+         !document.HasMember("url") || !document.HasMember("success") ||
+         !document.HasMember("fail") || !document.HasMember("progress"))
+      {
+         FailPromise(
+            UploadOperationCompleted::Result::UnexpetedResponse,
+            uploadPolicyJSON);
+
+         return;
+      }
+
+      auto form = std::make_unique<MultipartData>();
+
+      if (document.HasMember("fields"))
+      {
+         const auto& fields = document["fields"];
+
+         for (auto it = fields.MemberBegin(); it != fields.MemberEnd(); ++it)
+            form->Add(it->name.GetString(), it->value.GetString());
+      }
+
+      const auto fileField =
+         document.HasMember("field") ? document["field"].GetString() : "file";
+
+      const wxFileName name { mFileName };
+      form->AddFile(fileField, DeduceMimeType(name.GetExt()), name);
```

Comment:
I checked to see what would happen if a nonexisted file was given.

`AddFile` throws a `SimpleMessageBoxException`.

But we get here on a worker thread, right?  Is it safe to do that?  There will be a `CallAfter` that pops up the dialog, no problem, but is there a `catch` above us here?

If not, should there be a `GuardedCall`?

Do we need to guard any other calls into `lib-network-manager`?

Reply:
The file existence is checked on the main thread first. But I will improve safety here anyway.

Reply:
It's good for a class at a lower level not to make too-strong assumptions, or else, to make those assumptions explicit in comments in the header.

Reply:
This is private class. The file existence is checked by a public "factory" class before the object of this class is  created. There is a subtle chance that the file will be deleted while we are waiting to get here, so I will improve the exception guarantees at this specific point.

Reply:
This class is not even listed in header, to be fair.